### DISCUSSION TITLE
Deprecate current_power_w and today_energy_kwh switch properties

### DISCUSSION THREAD

**Discussion by thecode** [home-assistant/architecture#579] [Entity Models] [answered]
As per [Switch Entity properties](https://developers.home-assistant.io/docs/core/entity/switch/#properties) switch entity has two properties which are sensor related data:  `current_power_w ` and `today_energy_kwh`. These properties are better expressed as sensors rather then properties of a switch, making them a sensor will allow users to enable/disable, rename and easier automations based on the sensor value. Lovelace and device card support is also better using sensors. 

Sensor data is usually having higher frequency updates, in case of power the state of the sensor may be change in short intervals in terms of seconds, making it part of the switch entity will cause the switch state to be updated in higher frequency while the switch state may be static for a long period. Sensors also support state class which will allow better history for the data: [Sensor Entity state classes](https://developers.home-assistant.io/docs/core/entity/sensor/#available-state-classes)

There are currently 19 uses for `current_power_w ` and 10 uses `today_energy_kwh` in current codebase. 
I am suggesting to deprecate these properties. If we agree to remove I will migrate the integrations I can to use sensors.

**Comment by thecode** [comment, marked as answer]
@frenck @emontnemery What do you think?

**Reply by frenck** [reply]
I agree these should go, they don't match the rest of our design model anymore.

**Reply by balloob** [reply]
I agree.

**Comment by balloob** [comment]
When migrating the `current_power_w` to sensors, make sure to add `state_class: measurement` so that we will start compiling statistics for it.

**Comment by thecode** [comment]
I created a PR to update the documentation: https://github.com/home-assistant/developers.home-assistant/pull/982
I have started looking at changing the integrations and will also try to contact some of the code owners, but I don't use any of them and will be hard for me to validate I don't break them.
I suggest to create an issue and list the 20 integrations in the issue and update the issue when integrations are migrated to use sensors.

**Reply by thecode** [reply]
Demo platform update: https://github.com/home-assistant/core/pull/52424

**Comment by balloob** [comment]
I have created issues for all integrations that currently use one of the two electricity related attributes: https://github.com/home-assistant/core/issues?q=is%3Aissue+is%3Aopen+%22switch+energy+attributes+into+sensors%22

**Comment by mccasian** [comment]
We use now the rest sensor workaround. This works, but since the mystrom switches are WIFI switches, the readings result in Timeouts from time to time, and some automations with numeric trigger will be triggered after one bad + one good reading.
Is there a better workaround, or could you extend the mystrom integration and add mystrom sensors? The reading of power consumption from switch properties was much more reliable as you did it in mystrom integration

**Reply by MartinHjelmare** [reply]
If an integration has a problem it will need to be solved in that integration. There's nothing on an architecture level that stops an integration from exposing different entity types from the same device and coordinating the communication with the device centrally.

The architecture discussions forum is not a place for feature requests. Please use our community forum for that:

https://www.home-assistant.io/help/#bugs-feature-requests-and-alike

Thanks!
